api/src/main/java/io/grpc/CallOptions.java

```diff
@@ -427,19 +496,20 @@ public Integer getMaxOutboundMessageSize() {
   }
 
   /**
-   * Copy constructor.
+   * Copy CallOptions.
    */
-  private CallOptions(CallOptions other) {
-    deadline = other.deadline;
-    authority = other.authority;
-    credentials = other.credentials;
-    executor = other.executor;
-    compressorName = other.compressorName;
-    customOptions = other.customOptions;
-    waitForReady = other.waitForReady;
-    maxInboundMessageSize = other.maxInboundMessageSize;
-    maxOutboundMessageSize = other.maxOutboundMessageSize;
-    streamTracerFactories = other.streamTracerFactories;
+  private Builder fullBuild(CallOptions other) {
```

Comment:
- why is this method non-static?
- instead of `fullBuild` the canonical pattern is to define `toBuilder()` as an instance method so all callers of `fullBuild` will instead call `toBuilder()` on `other`.

Reply:
When I tested the performance of the compiler to optimize away the Builder allocation, I used `new Builder(this)` in each of the methods. That way the allocation never escaped from the method. Having a `toBuilder()` should still be possible to optimize away the allocation, but I think it becomes more dependent on inlining. I suggest we move this to Builder's constructor and explicitly call `new` in each `with` method.

Part of the trouble here is Java's escape analysis is powerful, but varies highly on the whims of the JIT. We also want Dalvik/ART to work well for Android. So the least amount of guessing and making what we are doing clear is helpful.

Reply:
> ...but I think it becomes more dependent on inlining. ...

Interesting. From this [article](https://blogs.oracle.com/javamagazine/post/escape-analysis-in-the-hotspot-jit-compiler): 

"EA provides an interesting insight: it is a feature that depends upon another optimization (automatic inlining) and is essentially useless without it."  

so I thought dependency on inlining is inherent.

Reply:
> > * why is this method non-static?
> 
> Thanks for your review. I want to know what the benefit is?

If a method doesn't use any instance data it should be marked static - so its callable from more sites. Even otherwise it is a good practice.